# Gene regulatory networks of two yeast species in multiple environments
### Data from Krieger et al. 2020 Genome Research and Fay et al. 2023 Genome Biology and Evolution
### Analysis inspired by Langfelder, Horvath, and Zhang's Weighted Gene Co-Expression Network Analysis R package (WGCNA)

Goal: Compare gene-gene co-expression networks between two species of yeast, *S. cerevisiae* and *S. paradoxus*

## R scripts (main folder)
Scripts belong to 3 categories:

#### 1) Data cleaning and analysis
- loading external datasets, clustering by expression dynamics in each environment, and running differential expression analysis
- ex) clean\_data.R and clustering.R scripts

#### 2) Data and Functions for figure scripts
- two scripts streamlining all .RData files and R functions to be used in figure generating scripts
- functions\_for\_figure\_scripts.R and data\_for\_figure\_scripts.R
  
#### 3) Figure generation scripts
- produces figures used in main publication as well as supplementary figures
- accepts .RData files from 
- ex) Fig\_Environmental\_Patterns.R and Fig\_Saturated\_Growth.R

## Data files
Also 3 categories:

#### 1) .RData files generated in R scripts
- ex) CorrelationClustering3Disp.RData
  
#### 2) slurm batch scripts
- used to align RNAseq reads on slurm-managed computing cluster

#### 3) External data files
- downloaded files from outside sources, unmodified to allow for reproducability

## Archive
All scripts no longer being used in analyses, potentially containing some useful code snippets for reference

Note: All .RData, files are un-tracked so they won't appear on github but can be generated by running the R scripts
